GVFS/GVFS.Common/ProductUpgrader.cs

```diff
@@ -83,37 +87,71 @@ protected ProductUpgrader()
             out ProductUpgrader newUpgrader,
             out string error)
         {
-            // Prefer to use the NuGet upgrader if it is configured. If the NuGet upgrader is not configured,
-            // then try to use the GitHubUpgrader.
-            if (NuGetUpgrader.TryCreate(tracer, fileSystem, gvfsConfig, credentialStore, dryRun, noVerify, out NuGetUpgrader nuGetUpgrader, out bool isConfigured, out error))
+            Dictionary<string, string> entries;
+            if (!gvfsConfig.TryGetAllConfig(out entries, out error))
             {
-                // We were successfully able to load a NuGetUpgrader - use that.
-                newUpgrader = nuGetUpgrader;
-                return true;
+                newUpgrader = null;
+                return false;
+            }
+
+            bool containsUpgradeFeedUrl = entries.ContainsKey(GVFSConstants.LocalGVFSConfig.UpgradeFeedUrl);
+            bool containsUpgradePackageName = entries.ContainsKey(GVFSConstants.LocalGVFSConfig.UpgradeFeedPackageName);
+            bool containsOrgInfoServerUrl = entries.ContainsKey(GVFSConstants.LocalGVFSConfig.OrgInfoServerUrl);
+
+            if (containsUpgradeFeedUrl || containsUpgradePackageName)
+            {
+                // We are configured for NuGet - determine if we are using ModernNuGetUpgrader or not
+                if (containsOrgInfoServerUrl)
+                {
+                    if (ModernNuGetUpgrader.TryCreate(
+                        tracer,
+                        fileSystem,
+                        gvfsConfig,
+                        new HttpClient(),
+                        credentialStore,
+                        dryRun,
+                        noVerify,
+                        out ModernNuGetUpgrader modernNuGetUpgrader,
+                        out error))
+                    {
+                        // We were successfully able to load a NuGetUpgrader - use that.
+                        newUpgrader = modernNuGetUpgrader;
+                        return true;
+                    }
+                    else
+                    {
+                        tracer.RelatedError($"{nameof(TryCreateUpgrader)}: Could not create upgrader. {error}");
+                        newUpgrader = null;

```

Comment:
Should this fallback to the NugetUpgrader or since there was the OrgInfo key we want to fail?

Reply:
The more I think about it - I think the current NuGetUpgrader will go away (either with this PR or soon after), and `ModernNuGetUpgrader` will just become `NuGetUpgrader`.

I do want to fail, and not fall back. In the case where the user "intends" to use one version of the upgrader, I didn't want to fall back to a different version "unexpectedly... and, I expect the fallback option to go away :)